### PR TITLE
[FLINK-14581][python] Let python UDF execution no longer rely on the flink directory structure to support running python UDFs on yarn.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request makes python UDF execution no longer rely on the flink directory structure to support running python UDFs on yarn.*

## Brief change log

  - *add `pyflink-udf-runner.sh`, `pyflink.zip`, `py4j-0.10.8.1-src.zip` and `cloudpickle-1.2.2-src.zip` as resources to `flink-python-{VERSION}.jar`.*
  - *add `RunnerEnvUtil` to extract above files before running python udf worker.*
  - *rework `pyflink-udf-runner.sh` and relevant tests to adapt to the above changes.*

## Verifying this change

This change is already covered by existing tests, such as *RunnerEnvUtilTest*, *test_process_mode_boot.py* etc.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/python/util/RunnerEnvUtil.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+public class RunnerEnvUtil {
+
+	public static final String[] PYTHON_BASIC_DEPENDENCIES = new String[] {
+		"pyflink.zip",
+		"py4j-0.10.8.1-src.zip",
+		"cloudpickle-1.2.2-src.zip",
+		"pyflink-udf-runner.sh"
+	};
+
+	public static List<File> extractBasicDependenciesFromResource(
```
Comment (hequn8128) [inline flink-python/src/main/java/org/apache/flink/python/util/RunnerEnvUtil.java]: Be careful with the format. You can take a look at the methods in `KeyedStream`. 

```
	public static List<File> extractBasicDependenciesFromResource(
			String tmpdir,
			ClassLoader classLoader,
			String prefix) throws IOException {
```

Also check other places.

File: flink-python/src/main/java/org/apache/flink/python/util/RunnerEnvUtil.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Utils for building the most basic environment for running python udf workers.
+ * The basic environment does not include python part of Apache Beam.
+ * Users need to prepare it themselves.
+ */
+public class RunnerEnvUtil {
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/python/util/RunnerEnvUtil.java]: Due to the extract method is reused by the `PythonResourceExtractor`, I only move the `appendEnvironmentVariable` into AbstractPythonFunctionRunner.

File: flink-python/src/main/java/org/apache/flink/python/util/RunnerEnvUtil.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			} else if (systemEnv.get(ConfigConstants.ENV_FLINK_LIB_DIR) != null) {
+				// log directory of standalone mode
+				File flinkHomeDir = new File(systemEnv.get(ConfigConstants.ENV_FLINK_LIB_DIR)).getParentFile();
+				result.put("FLINK_LOG_DIR", new File(flinkHomeDir, "log").getAbsolutePath());
+			} else if (systemEnv.get(ConfigConstants.ENV_FLINK_HOME_DIR) != null) {
+				// log directory of pyflink shell mode or user specified flink
+				result.put("FLINK_LOG_DIR",
+					new File(systemEnv.get(ConfigConstants.ENV_FLINK_HOME_DIR), "log").getAbsolutePath());
+			}
+		}
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/python/util/RunnerEnvUtil.java]: Currently using `System.getProperty("log.file")` (default value of the log path of taskmanager) here, after moving the environment creation into EnvironmentManager in [#10017](https://github.com/apache/flink/pull/10017) we can pass the true log path to this method easily.

**Comment by WeiZhong94** [comment]
@hequn8128 Thanks for your review! I have addressed your comments in the latest commit, please take a look again :).

**Comment by WeiZhong94** [comment]
@hequn8128 Thanks for your review again! I have updated this PR. Please take a look. :)

**Comment by WeiZhong94** [comment]
@hequn8128 Thanks for your reminder, I have fixed the test failure in the latest commit.

**Comment by hequn8128** [comment]
@WeiZhong94 Thanks a lot for the update. Merging...

**Pull request closed without merging**